Codegen python bindings to access attributes of grad_fn

soulitzer opened the pull request:
Fixes #9922

Adds python bindings to *selected* fields that grad_fn saves - we did not add python bindings to certain types such as 'TypeAndSize' and 'TensorGeometry'. All field names are prefixed with `_saved_` so they are easy to discern. User code should not depend on particular saved fields to exist as what grad_fn saves for the backward pass is considered an implementation detail and thus prone to change.

Warning: Not all parameters that are passed in are necessarily stored to be used for the backward pass. What you put in is not necessarily what you get out either. Here we pass `kernel_size=3`, but `b.grad_fn._saved_kernel_size` returns `(3, 3)` instead of 3. It seems to vary case-by-case.

For example:
```
import torch
import torch.nn as nn

model = nn.Conv2d(in_channels=3, out_channels=32, kernel_size=3, stride=2, padding=1, dilation=1)

a = torch.ones(1, 3, 32, 32, requires_grad=True)
b = model(a)

print("kernel_size: ", b.grad_fn._saved_kernel_size)
print("stride: ", b.grad_fn._saved_stride) # returns tuple: (3, 3)
# print("dilation: ", b.grad_fn._saved_dilation) # dilation is not stored for backward pass
print("padding: ", b.grad_fn._saved_padding)
print("weight: ", b.grad_fn._saved_weight)
```

Sample of generated code:
```
PyObject* THPThnnConv2DBackward_self_getter(THPCppFunction *self, void *_unused) {
  const auto& prop = static_cast<ThnnConv2DBackward*>(self->cdata.get())->self_;
  return THPVariable_Wrap(prop.unpack());
}

PyObject* THPThnnConv2DBackward_weight_getter(THPCppFunction *self, void *_unused) {
  const auto& prop = static_cast<ThnnConv2DBackward*>(self->cdata.get())->weight_;
  return THPVariable_Wrap(prop.unpack());
}

PyObject* THPThnnConv2DBackward_kernel_size_getter(THPCppFunction *self, void *_unused) {
  auto prop = static_cast<ThnnConv2DBackward*>(self->cdata.get())->kernel_size;
  PyObject* tup = PyTuple_New((Py_ssize_t) prop.size());
  for (int i = 0; i < prop.size(); i++) {
    PyTuple_SetItem(tup, (Py_ssize_t) i, PyLong_FromUnsignedLong((uint64_t) prop[i]));
  }
  return tup;
}

PyObject* THPThnnConv2DBackward_stride_getter(THPCppFunction *self, void *_unused) {
  auto prop = static_cast<ThnnConv2DBackward*>(self->cdata.get())->stride;
  PyObject* tup = PyTuple_New((Py_ssize_t) prop.size());
  for (int i = 0; i < prop.size(); i++) {
    PyTuple_SetItem(tup, (Py_ssize_t) i, PyLong_FromUnsignedLong((uint64_t) prop[i]));
  }
  return tup;
}

PyObject* THPThnnConv2DBackward_padding_getter(THPCppFunction *self, void *_unused) {
  auto prop = static_cast<ThnnConv2DBackward*>(self->cdata.get())->padding;
  PyObject* tup = PyTuple_New((Py_ssize_t) prop.size());
  for (int i = 0; i < prop.size(); i++) {
    PyTuple_SetItem(tup, (Py_ssize_t) i, PyLong_FromUnsignedLong((uint64_t) prop[i]));
  }
  return tup;
}

PyObject* THPThnnConv2DBackward_finput_getter(THPCppFunction *self, void *_unused) {
  const auto& prop = static_cast<ThnnConv2DBackward*>(self->cdata.get())->finput_;
  return THPVariable_Wrap(prop.unpack());
}

PyObject* THPThnnConv2DBackward_fgrad_input_getter(THPCppFunction *self, void *_unused) {
  const auto& prop = static_cast<ThnnConv2DBackward*>(self->cdata.get())->fgrad_input_;
  return THPVariable_Wrap(prop.unpack());
}

static struct PyGetSetDef ThnnConv2DBackward_properties[] = {
  THP_FUNCTION_DEFAULT_PROPERTIES,
  {(char*)"_saved_self", (getter)THPThnnConv2DBackward_self_getter, nullptr, nullptr, nullptr},
  {(char*)"_saved_weight", (getter)THPThnnConv2DBackward_weight_getter, nullptr, nullptr, nullptr},
  {(char*)"_saved_kernel_size", (getter)THPThnnConv2DBackward_kernel_size_getter, nullptr, nullptr, nullptr},
  {(char*)"_saved_stride", (getter)THPThnnConv2DBackward_stride_getter, nullptr, nullptr, nullptr},
  {(char*)"_saved_padding", (getter)THPThnnConv2DBackward_padding_getter, nullptr, nullptr, nullptr},
  {(char*)"_saved_finput", (getter)THPThnnConv2DBackward_finput_getter, nullptr, nullptr, nullptr},
  {(char*)"_saved_fgrad_input", (getter)THPThnnConv2DBackward_fgrad_input_getter, nullptr, nullptr, nullptr},
  {nullptr} /* sentinel */
};

...

void initialize_autogenerated_functions() {
...
  static PyTypeObject ThnnConv2DBackwardClass;
  addClass<ThnnConv2DBackward>(ThnnConv2DBackwardClass, "ThnnConv2DBackward", ThnnConv2DBackward_properties);
...
}
```

Before:
```
void initialize_autogenerated_functions() {
...
  static PyTypeObject ThnnConv2DBackwardClass;
  addClass<ThnnConv2DBackward>(ThnnConv2DBackwardClass, "ThnnConv2DBackward");
...
}
```

soulitzer committed:
Initial changes

soulitzer committed:
Refactor

soulitzer committed:
Support vec of savedvar

albanD commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -82,6 +77,91 @@
+# This generates the definitions for:
+#   (1) The PyTypeObject for each backward grad_fn subclassing Node
+#   (2) The entry for PyTypeObject's tp_getset slot (an array of PyGetSetDef structs)
+#     We generate one PyGetSetDef struct for each of grad_fn's saved inputs and outputs
```
nit: alignment is wrong no?

albanD commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -150,13 +234,18 @@ def process_function(info: DifferentiabilityInfo, template: CodeTemplate) -> str
         name = var.name
+        should_append_getsetdef = True
+
+        get_prop = GET_PROP.substitute(op=info.op, name=name)
```
this one is used only once no? So it can be moved down

albanD commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -150,13 +234,18 @@ def process_function(info: DifferentiabilityInfo, template: CodeTemplate) -> str
+        should_append_getsetdef = True
+
+        get_prop = GET_PROP.substitute(op=info.op, name=name)
+        get_prop_optionalarray = GET_PROP_OPTIONAL_ARRAY.substitute(op=info.op, name=name)
```
while this is used twice, I think it is fine to move it down to simplify readability.

albanD reviewed (commented):
As discussed offline, these wouldn't be stable in any way so we want to add `_` before them to properly mark they are internal.

It won't be useful for general users but can be very useful for debugging and also utility tools like torchviz or anomaly mode to improve the debugging.

soulitzer committed:
Add more types and refactor and tests

soulitzer commented:
Added bindings for some more types that I had missed previously.This led to a refactor that simplified the code a bit as well. Also added tests.

There are still a couple types such as `TypeAndSize`, `TensorType`, and `TensorGeometry` that I don't see a clear way of providing python bindings for, but since the main use for this is still for debugging purposes (i.e., torchviz), I think its okay to leave those out for now. Its going to be pretty straightforward anyway to add them later if someone finds a compelling use case.

soulitzer committed:
simplify

albanD commented on test/test_autograd.py:
```diff
@@ -4282,6 +4282,76 @@ def manual_increase_gradient(grad):
+        b = torch.ones(1, requires_grad=True)
+        out = torch.stack([a, b], dim=0)
+        self.assertEqual(out.grad_fn._tensors, (a, b))              # TensorList -> Tuple[Tensor]
+        self.assertIsInstance(out.grad_fn._tensors, tuple)
```
The assertEqual above should already be checking for the type and content no?
Same for al lthe tests, you can remove the duplicate checks (if any)

soulitzer replied:
Removed the tuple ones. the assertEqual doesn't check types are the same for numeric types though

albanD commented on test/test_autograd.py:
```diff
@@ -4282,6 +4282,76 @@ def manual_increase_gradient(grad):
 
+    def test_grad_fn_attr_bindings(self):
+        # Check that the getter of each type returns what we want
+        # See `gen_autograd_functions.py` for how getters are generated
```
Can you add a comment here that future users can safely update the tests in the future if they change any of these functions implementations.
This test is just to test the binding and not that the attributes are always present for these functions.

albanD commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -82,6 +77,150 @@
+
+GETTER_BODY_SCALAR = """\
+if (prop.isComplex()) {
+  auto cprop = prop.to<c10::complex<double>>();
```
Why are these always double? What happens if they are float32/float16?

soulitzer replied:
Added a test for this for the float32 case

albanD commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -82,6 +77,150 @@
+""")
+
+PY_GETSETDEF_STRUCT = CodeTemplate("""\
+{(char*)"_${name}", (getter)THP${op}_${name}_getter, nullptr, nullptr, nullptr}""")
```
Since one of the goal here is to allow the users to easily find these irrespective of their name, maybe we should consider giving them a common prefix like: `_saved_foo`. This will allow users to use `dir(Node)` and filter anything that starts with `_saved_` to get all the attributes they want.
What do you think?

soulitzer committed:
Address comments

albanD reviewed (approved):
Looks good to me, thanks for the update!
